File: C-Sharp/CodilityScenarios/4. Counting Elements/MaxCounters.cs

namespace CodilityScenarios
{
    /// <summary>
    /// Write a function that, given an integer N and a non-empty zero-indexed 
    /// array A consisting of M integers, returns a sequence of integers 
    /// representing the values of the counters.
    /// </summary>
    public class MaxCounters
    {
        public int[] solution(int counterLimit, int[] numericArray)
        {
            // Instantiate the counters
            int[] counters = new int[counterLimit];
            int maxValue = 0;

            for (int i = 0; i < numericArray.Length; i++)
            {
                int element = numericArray[i];

                // Guard against out of range 
                if (element > counterLimit)
                {
                    // check if we need to max counters
                    if (element == (counterLimit + 1))
                    {
                        // Dot Net Core Solution
                        // Array.Fill(counters, counters.Max());

                        // .NET 4.5 :(
                        for (int ii = 0; ii < counterLimit; ii++){
                            
                            counters[ii] = maxValue;
                        }
                    }

                    continue;
                }

                // Standard Increment    
                counters[element - 1]++;

                // Track the max value
                if(maxValue < counters[element - 1])
                {
                    maxValue = counters[element - 1];
                }
            }

            return counters;
        }
    }
}

File: Python/4. Counting Elements/MaxCounters.py
# Write a function that, given an integer N and a non-empty zero-indexed 
# array A consisting of M integers, returns a sequence of integers 
# representing the values of the counters.
class MaxCounters(object):

    def solution(self, counterLimit, numericArray):

        # create an array of zeros
        counters = [0] * counterLimit
        maxValue = 0

        # iterate on the array
        for value in numericArray:

            #check for array out of bounds
            if(value > counterLimit):

                if value == (counterLimit + 1):
                    counters = [maxValue] * counterLimit
                continue
            # assign the counter
            counters[value -1] = (counters[value -1] + 1)
            
            # track the max value
            if  maxValue < counters[value -1]:
                maxValue = counters[value -1]

        return counters